ctap2token/example/main.go

```diff
@@ -0,0 +1,191 @@
+	"os"
+	"strings"
+
+	"github.com/flynn/u2f/ctap2token"
+	"github.com/flynn/u2f/u2fhid"
+	"github.com/grantae/certinfo"
+)
+
+func main() {
+	devices, err := u2fhid.Devices()
+	if err != nil {
+		panic(err)
+	}
+
+	for _, d := range devices {
+		dev, err := u2fhid.Open(d)
+		if err != nil {
+			panic(err)
+		}
+
+		token := ctap2token.NewToken(dev)
+
+		infos, err := token.GetInfo()
+		if err != nil {
+			panic(err)
+		}
+		fmt.Printf("Token infos:\n%#v\n", infos)
+
+		// https://fidoalliance.org/specs/fido-v2.0-ps-20190130/fido-client-to-authenticator-protocol-v2.0-ps-20190130.html#gettingSharedSecret
+		fmt.Println("Retrieving key agreement from authenticator")
+		cp1, err := token.ClientPIN(&ctap2token.ClientPINRequest{
+			PinProtocol: ctap2token.PinProtoV1,
+			SubCommand:  ctap2token.GetKeyAgreement,
+		})
+		if err != nil {
+			panic(err)
+		}
+
+		fmt.Println("Generating platform key pair")
+		b, bGX, bGY, err := elliptic.GenerateKey(elliptic.P256(), rand.Reader)
+		if err != nil {
+			panic(err)
+		}
+
+		aG := cp1.KeyAgreement
+		aGX := new(big.Int)
+		aGX.SetBytes(aG.X)
+
+		aGY := new(big.Int)
+		aGY.SetBytes(aG.Y)
+
+		rX, _ := elliptic.P256().ScalarMult(aGX, aGY, b)
+
+		h := sha256.New()
+		_, err = h.Write(rX.Bytes())
+		if err != nil {
+			panic(err)
+		}
+
+		sharedSecret := h.Sum(nil)
+		fmt.Println("Generated shared secret")
+
+		reader := bufio.NewReader(os.Stdin)
+		fmt.Print("Enter PIN: ")
+		userPIN, _ := reader.ReadString('\n')
+
+		h.Reset()
+		_, err = h.Write([]byte(strings.TrimSpace(userPIN)))
+		if err != nil {
+			panic(err)
+		}
+
+		pinHash := h.Sum(nil)
+		pinHash = pinHash[:aes.BlockSize]
+
+		pinHashEnc := make([]byte, aes.BlockSize)
+		c, err := aes.NewCipher(sharedSecret)
+		if err != nil {
+			panic(err)
+		}
+		iv := make([]byte, aes.BlockSize)
+		cbcEnc := cipher.NewCBCEncrypter(c, iv)
+		cbcEnc.CryptBlocks(pinHashEnc, pinHash)
+		fmt.Println("Encrypted user PIN using shared secret")
+
+		pinResp, err := token.ClientPIN(&ctap2token.ClientPINRequest{
+			SubCommand: ctap2token.GetPinToken,
+			KeyAgreement: &ctap2token.COSEKey{
+				X:       bGX.Bytes(),
+				Y:       bGY.Bytes(),
+				KeyType: ctap2token.EC2,            // not required ?
+				Curve:   ctap2token.P256,           // not required ?
+				Alg:     ctap2token.ECDHES_HKDF256, // not required ?
+			},
+			PinHashEnc:  pinHashEnc,
+			PinProtocol: ctap2token.PinProtoV1,
+		})
+		if err != nil {
+			panic(err)
+		}
+
+		// Decrypt pinToken using shared secret
+		pinHashDec := make([]byte, aes.BlockSize)
+		cbcDec := cipher.NewCBCDecrypter(c, iv)
+		cbcDec.CryptBlocks(pinHashDec, pinResp.PinToken)
+		fmt.Println("Decrypted authenticator pinToken")
+
+		clientDataHash := make([]byte, 32)
+		if _, err := rand.Read(clientDataHash); err != nil {
+			panic(err)
+		}
+
+		userID := make([]byte, 32)
+		if _, err := rand.Read(userID); err != nil {
+			panic(err)
+		}
+
+		mac := hmac.New(sha256.New, pinHashDec)
+		_, err = mac.Write(clientDataHash)
```

Comment:
The host side of the PIN protocol should be implemented in a separate package instead of inline in the example.

Reply:
:+1: Moved all of it under a `pin` package